File: app/syncthing/base_api.py
import os
import json
import logging
import requests

from requests import Response

from .syncthing_exception import SyncthingException


DEFAULT_TIMEOUT = 10

class BaseAPI:
    """ Placeholder for HTTP REST API URL prefix. """

    prefix = ''

    def __init__(
        self,
        api_key: str,
        host: str = 'localhost',
        port: int = 8384,
        timeout: int = DEFAULT_TIMEOUT,
        is_https: bool = False,
        ssl_cert_file: str | None = None
    ):
        if ssl_cert_file:
            if not os.path.exists(ssl_cert_file):
                raise SyncthingException(f'ssl_cert_file does not exist at location, {ssl_cert_file}')

        self.api_key = api_key
        self.host = host
        self.port = port
        self.timeout = timeout
        self.is_https = is_https
        self.ssl_cert_file = ssl_cert_file
        self.verify = True if ssl_cert_file or is_https else False
        self._headers = {
            'X-API-Key': api_key
        }
        self.proto: str = 'https' if is_https else 'http'
        self.url = f'{self.proto}://{self.host}:{self.port}'
        self.logger = logging.getLogger(__name__)

    def get(
        self,
        endpoint: str,
        data: dict | str | None = None,
        headers: dict | None = None,
        params: dict | None = None,
        return_response: bool = False,
        raw_exceptions: bool = False
    ) -> Response | int | str | dict | list:
        return self._request('GET', self.prefix + endpoint, data, headers, params, return_response, raw_exceptions)

    def post(
        self,
        endpoint: str,
        data: dict | str | None = None,
        headers: dict | None = None,
        params: dict | None = None,
        return_response: bool = False,
        raw_exceptions: bool = False
    ) -> Response | int | str | dict | list:
        return self._request('POST', self.prefix + endpoint, data, headers, params, return_response, raw_exceptions)

    def put(
        self,
        endpoint: str,
        data: dict | str | None = None,
        headers: dict | None = None,
        params: dict | None = None,
        return_response: bool = False,
        raw_exceptions: bool = False
    ) -> Response | int | str | dict | list:
        return self._request('PUT', self.prefix + endpoint, data, headers, params, return_response, raw_exceptions)

    def delete(
        self,
        endpoint: str,
        data: dict | str | None = None,
        headers: dict | None = None,
        params: dict | None = None,
        return_response: bool = False,
        raw_exceptions: bool = False
    ) -> Response | int | str | dict | list:
        return self._request('DELETE', self.prefix + endpoint, data, headers, params, return_response, raw_exceptions)

    def _request(
        self,
        method: str,
        endpoint: str,
        data: dict | str | None = None,
        headers: dict | None = None,
        params: dict | None = None,
        return_response: bool = False,
        raw_exceptions: bool = False
    ) -> Response | int | str | dict | list:
        method = method.upper()

        endpoint = self.url + endpoint

        if method not in ('GET', 'POST', 'PUT', 'DELETE'):
            raise SyncthingException(f'unsupported http verb requested, {method}')

        if data is None:
            data = {}

        if headers is None:
            headers = {}

        try:
            if not isinstance(data, dict):
                raise AssertionError()
            if not isinstance(headers, dict):
                raise AssertionError()

            headers.update(self._headers)

            response = requests.request(
                method,
                endpoint,
                data=json.dumps(data),
                params=params,
                timeout=self.timeout,
                cert=self.ssl_cert_file,
                headers=headers
            )

            if not return_response:
                response.raise_for_status()

        except requests.RequestException as e:
            if raw_exceptions:
                raise e
            raise SyncthingException('http request error') from e

        else:
            if return_response:
                return response

            if response.status_code != requests.codes.ok:
                self.logger.error(f'{response.status_code} {response.reason} ({response.url}): {response.text}')
                return response

            if 'json' in response.headers.get('Content-Type', 'text/plain').lower():
                json_data = response.json()
            else:
                content = response.content.decode('utf-8')
                if content and content[0] == '{' and content[-1] == '}':
                    json_data = json.loads(content)
                else:
                    return content

            if isinstance(json_data, dict) and json_data.get('error'):
                api_err = json_data.get('error')
                raise SyncthingException(api_err)
            return json_data

__all__ = [
    'DEFAULT_TIMEOUT',
    'BaseAPI'
]

File: app/src/syncthing/BaseAPI.ts
import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import * as fs from 'fs';
import * as https from 'https';

import SyncthingException from './SyncthingException';
import ServiceConfig from './ServiceConfig';

export interface RequestData {
  device?: string;
  folder?: string | string[];
  ignore?: string[];
  id?: string;
  length?: number;
  message?: string;
}

export interface RequestHeaders {
  'X-API-Key'?: string;
  'Accept-Language'?: string;
}

export interface RequestParameters {
  device?: string;
  folder?: string;
  file?: string;
  levels?: number;
  prefix?: string;
  page?: number;
  perpage?: number;
  sub?: string;
  next?: string;
  since?: number;
  limit?: number;
  timeout?: number;
  events?: string;
  address?: string;
  current?: string;
}


export default class BaseAPI {
  // Placeholder for HTTP REST API URL prefix.
  protected prefix: string = '';

  protected serviceConfig: ServiceConfig;
  protected verify: boolean;
  protected headers: RequestHeaders;
  protected protocol: string;
  protected url: string;
  protected logger: Console;

  constructor(config: ServiceConfig, logger: Console) {
    if (config.sslCertFile && !fs.existsSync(config.sslCertFile)) {
      throw new SyncthingException(`ssl_cert_file does not exist at location, ${config.sslCertFile}`);
    }

    this.serviceConfig = config;
    this.verify = !!(config.sslCertFile || config.isHttps);
    this.headers = {
      'X-API-Key': config.apiKey
    };
    this.protocol = config.isHttps ? 'https' : 'http';
    this.url = `${this.protocol}://${config.host}:${config.port}`;
    this.logger = logger;
  }

  async get<T>(
    endpoint: string,
    data?: RequestData,
    headers?: RequestHeaders,
    params?: RequestParameters
  ): Promise<T> {
    return this.request<T>('GET', this.prefix + endpoint, data, headers, params);
  }

  async post<T>(
    endpoint: string,
    data?: RequestData,
    headers?: RequestHeaders,
    params?: RequestParameters
  ): Promise<T> {
    return this.request<T>('POST', this.prefix + endpoint, data, headers, params);
  }

  async put<T>(
    endpoint: string,
    data?: RequestData,
    headers?: RequestHeaders,
    params?: RequestParameters
  ): Promise<T> {
    return this.request<T>('PUT', this.prefix + endpoint, data, headers, params);
  }

  async delete<T>(
    endpoint: string,
    data?: RequestData,
    headers?: RequestHeaders,
    params?: RequestParameters
  ): Promise<T> {
    return await this.request<T>('DELETE', this.prefix + endpoint, data, headers, params);
  }

  protected async request<T>(
    method: string,
    endpoint: string,
    data?: RequestData,
    headers?: RequestHeaders,
    params?: RequestParameters
  ): Promise<T> {
    const response: AxiosResponse<T> | undefined = await this.raw_request(method, endpoint, data, headers, params);
    if (!response) {
      throw new SyncthingException('No response from Syncthing API');
    }

    if (response && response.status >= 400) {
      throw new SyncthingException(`HTTP Error: ${response.status} ${response.statusText}`);
    }

    if (typeof response.data === 'object' && response.data !== null && 'error' in response.data && response.data.error !== null) {
      throw new SyncthingException(`Response contains the error: ${response.data.error}`);
    }

    return response.data;
  }

  protected async raw_request<T>(
    method: string,
    endpoint: string,
    data?: RequestData,
    headers?: RequestHeaders,
    params?: RequestParameters
  ): Promise<AxiosResponse<T>> {
    method = method.toUpperCase();
    endpoint = new URL(endpoint, this.url).toString();

    if (!['GET', 'POST', 'PUT', 'DELETE'].includes(method)) {
      throw new SyncthingException(`unsupported http verb requested, ${method}`);
    }

    if (data === undefined) {
      data = {};
    }

    if (headers === undefined) {
      headers = {};
    }

    if (typeof data !== 'object' || Array.isArray(data)) {
      throw new Error('Data must be an object');
    }
    if (typeof headers !== 'object' || Array.isArray(headers)) {
      throw new Error('Headers must be an object');
    }

    let result: AxiosResponse<T>;
    try {
      headers = Object.assign(headers, this.headers);

      const axiosConfig: AxiosRequestConfig = {
        url: endpoint,
        method: method,
        headers: headers as object,
        params: params,
        data: data,
        timeout: this.serviceConfig.timeout * 1000,
      };

      if (this.verify && this.serviceConfig.sslCertFile) {
        axiosConfig.httpsAgent = new https.Agent({
          ca: fs.readFileSync(this.serviceConfig.sslCertFile, 'utf8')
        });
      }
      result = await axios.request<T>(axiosConfig);
      /* eslint-disable @typescript-eslint/no-explicit-any */
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        throw new SyncthingException('HTTP request error', { cause: error });
      }
      result = error.response;
    }
    return result;
  }
}
